fastparquet/writer.py

```diff
@@ -607,6 +604,31 @@ def write(filename, data, partitions=[0], encoding="PLAIN",
                               open_with)
 
 
+def partition_on_columns(data, columns, root_path, partname, fmd, sep,
+                         compression, encoding, open_with, mkdirs):
+    """Split each data division on columns, and write in structured directories
+    """
+    gb = data.groupby(columns)
+    remaining = list(data)
+    for column in columns:
+        remaining.remove(column)
+    for key in gb.indices:
+        df = gb.get_group(key)[remaining]
+        path = sep.join(["%s=%s" % (name, val)
+                         for name, val in zip(columns, key)])
+        relname = sep.join([path, partname])
+        mkdirs(root_path + sep + path)
+        fullname = sep.join([root_path, path, partname])
+        with open_with(fullname) as f2:
+            rg = make_part_file(f2, df, fmd.schema,
+                                compression=compression,
+                                encoding=encoding)
+        if rg is not None:
+            for chunk in rg.columns:
+                chunk.file_path = relname
+            fmd.row_groups.append(rg)
```

Comment:
Of course we cannot cover everything that a user does, but `cut()` iexists so that they can make a column that does have the required cardinality, and use that for partitioning.